src/python/pants/backend/typescript/dependency_inference/rules.py

```diff
@@ -129,36 +114,45 @@ async def infer_typescript_source_dependencies(
     )
     metadata = await _prepare_inference_metadata(request.field_set.address, source.file_path)
 
-    import_strings = await Get(
-        NativeParsedJavascriptDependencies,
-        NativeDependenciesRequest(sources.snapshot.digest, metadata),
+    import_strings, candidate_pkgs = await concurrently(
+        parse_javascript_deps(NativeDependenciesRequest(sources.snapshot.digest, metadata)),
+        map_candidate_node_packages(
+            RequestNodePackagesCandidateMap(request.field_set.address), **implicitly()
+        ),
     )
 
-    owning_targets_collection = await MultiGet(
-        Get(Targets, TypeScriptFileImportPath, TypeScriptFileImportPath(path=path))
-        for path in import_strings.file_imports
-    )
-    owning_targets = [tgt for targets in owning_targets_collection for tgt in targets]
-
-    non_path_string_bases = FrozenOrderedSet(
-        non_path_string.partition(os.path.sep)[0]
-        for non_path_string in import_strings.package_imports
-    )
-
-    candidate_pkgs = await Get(
-        NodePackageCandidateMap, RequestNodePackagesCandidateMap(request.field_set.address)
+    imports = dict(
+        zip(
+            import_strings.imports,
+            await concurrently(
+                _determine_import_from_candidates(
+                    candidates,
+                    candidate_pkgs,
+                    file_extensions=TS_FILE_EXTENSIONS + JS_FILE_EXTENSIONS,
+                )
+                for string, candidates in import_strings.imports.items()
+            ),
+        )
     )
-
-    pkg_addresses = (
-        candidate_pkgs[pkg_name] for pkg_name in non_path_string_bases if pkg_name in candidate_pkgs
+    _handle_unowned_imports(
+        request.field_set.address,
+        nodejs_infer.unowned_dependency_behavior,
+        frozenset(
+            string
+            for string, addresses in imports.items()
+            if not addresses and not _is_node_builtin_module(string)
+        ),
     )
-
-    return InferredDependencies(
-        itertools.chain(
-            pkg_addresses,
-            (tgt.address for tgt in owning_targets if tgt.has_field(TypeScriptSourceField)),
-        )
+    _handle_unowned_imports(
+        request.field_set.address,
+        nodejs_infer.unowned_dependency_behavior,
+        frozenset(
+            string
+            for string, addresses in imports.items()
+            if not addresses and not _is_node_builtin_module(string)
+        ),
     )
```

Comment:
Is function call duplicated? Or am I missing a subtle difference?

Reply:
Oh yes, good catch. Fixed!